Unable to execute 'jx boot' from outside a boot repo

### Summary 

This is an side-effect of the changes merged for https://github.com/jenkins-x/jx/issues/6653

### Steps to reproduce the behavior

From an empty directory execute `jx boot`:

```bash
jx --verbose boot
Creating boot config with defaults, as not in an existing boot directory with a git repository.
hello <nil>
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x199 pc=0x305bdd0]

goroutine 1 [running]:
github.com/jenkins-x/jx/pkg/cmd/opts.(*CommonOptions).ConfigureCommonOptions(0xc000458000, 0x0, 0x0, 0x0)
	/Users/hardy/work/go/src/github.com/jenkins-x/jx/pkg/cmd/opts/team_settings.go:169 +0x90
github.com/jenkins-x/jx/pkg/cmd/boot.(*BootOptions).Run(0xc00069aa00, 0x0, 0x0)
	/Users/hardy/work/go/src/github.com/jenkins-x/jx/pkg/cmd/boot/boot.go:158 +0x1bb
github.com/jenkins-x/jx/pkg/cmd/boot.NewCmdBoot.func1(0xc00069d900, 0xc0004f7940, 0x0, 0x1)
	/Users/hardy/work/go/src/github.com/jenkins-x/jx/pkg/cmd/boot/boot.go:92 +0x6f
github.com/spf13/cobra.(*Command).execute(0xc00069d900, 0xc0004f7930, 0x1, 0x1, 0xc00069d900, 0xc0004f7930)
	/Users/hardy/work/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830 +0x2ae
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001c0000, 0xc0002c5c80, 0x66c5e20, 0xc0000b8000)
	/Users/hardy/work/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fc
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/hardy/work/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/jenkins-x/jx/cmd/jx/app.Run(0x0, 0x0, 0x0, 0xc000a37f88, 0xc00005c178)
	/Users/hardy/work/go/src/github.com/jenkins-x/jx/cmd/jx/app/jx.go:19 +0x185
main.main()
	/Users/hardy/work/go/src/github.com/jenkins-x/jx/cmd/jx/jx.go:11 +0x32
```

--- a/pkg/cmd/boot/boot.go
+++ b/pkg/cmd/boot/boot.go
@@ -111,8 +111,6 @@ func NewCmdBoot(commonOpts *opts.CommonOptions) *cobra.Command {
 
 // Run runs this command
 func (o *BootOptions) Run() error {
-	info := util.ColorInfo
-
 	err := o.verifyClusterConnection()
 	if err != nil {
 		return err
@@ -123,8 +121,13 @@ func (o *BootOptions) Run() error {
 	if o.AttemptRestore {
 		err := o.restoreFromDevEnvRepo()
 		if err != nil {
-			return err
+			return errors.Wrapf(err, "unable tp restore dev environment repo")
 		}
+	}
+
+	gitURL, gitRef := o.determineGitURLAndRef()
+	if gitURL == "" {
+		return util.MissingOption("git-url")
 	}
 
 	isBootClone, err := existingBootClone(o.Dir)
@@ -132,28 +135,28 @@ func (o *BootOptions) Run() error {
 		return errors.Wrapf(err, "failed to check if this is an existing boot clone")
 	}
 
-	gitURL, gitRef, err := gits.GetGitInfoFromDirectory(o.Dir, o.Git())
+	isGitRepo := o.isGitRepo(o.Dir)
+	if isGitRepo && !isBootClone {
+		return errors.Errorf("trying to execute 'jx boot' from a non requirements repo")
+	}
+
+	if !isBootClone {
+		gitInfo, err := gits.ParseGitURL(gitURL)
+		if err != nil {
+			return errors.Wrapf(err, "failed to parse git URL %s", gitURL)
+		}
+		dir := filepath.Join(o.Dir, gitInfo.Name)
+		cloneDir, err := o.createBootClone(gitURL, gitRef, dir)
+		if err != nil {
+			return errors.Wrapf(err, "unable to clone %s", gitURL)
+		}
+		o.Dir = cloneDir
+	}
+
+	requirements, requirementsFile, err := config.LoadRequirementsConfig(o.Dir)
 	if err != nil {
-		log.Logger().Info("Creating boot config with defaults, as not in an existing boot directory with a git repository.")
-		gitURL = config.DefaultBootRepository
-		gitRef = config.DefaultVersionsRef
+		return errors.Wrapf(err, "unable to load %s", config.RequirementsConfigFileName)
 	}
-
-	if o.GitURL != "" {
-		log.Logger().Infof("GitURL provided, overriding the current value: %s with %s", util.ColorInfo(gitURL), util.ColorInfo(o.GitURL))
-		gitURL = o.GitURL
-	}
-
-	if o.GitRef != "" {
-		log.Logger().Infof("GitRef provided, overriding the current value: %s with %s", util.ColorInfo(gitRef), util.ColorInfo(o.GitRef))
-		gitRef = o.GitRef
-	}
-
-	if gitURL == "" {
-		return util.MissingOption("git-url")
-	}
-
-	requirements, _, _ := config.LoadRequirementsConfig(o.Dir)
 
 	err = o.ConfigureCommonOptions(requirements)
 	if err != nil {
@@ -176,94 +179,6 @@ func (o *BootOptions) Run() error {
 		}
 	}
 
-	if !isBootClone {
-		log.Logger().Infof("No Jenkins X pipeline file %s or no jx boot requirements file %s found. You are not running this command from inside a "+
-			"Jenkins X Boot git clone", info(config.ProjectConfigFileName), info(config.RequirementsConfigFileName))
-
-		gitInfo, err := gits.ParseGitURL(gitURL)
-		if err != nil {
-			return errors.Wrapf(err, "failed to parse git URL %s", gitURL)
-		}
-
-		repo := gitInfo.Name
-		cloneDir := filepath.Join(o.Dir, repo)
-
-		if !o.BatchMode {
-			log.Logger().Infof("To continue we will clone %s @ %s to %s", info(gitURL), info(gitRef), info(cloneDir))
-
-			help := "A git clone of a Jenkins X Boot source repository is required for 'jx boot'"
-			message := "Do you want to clone the Jenkins X Boot Git repository?"
-			if answer, err := util.Confirm(message, true, help, o.GetIOFileHandles()); err != nil {
-				return err
-			} else if !answer {
-				return fmt.Errorf("Please run this command again inside a git clone from a Jenkins X Boot repository")
-			}
-		}
-
-		bootCloneExists, err := util.DirExists(cloneDir)
-		if err != nil {
-			return err
-		}
-		if bootCloneExists {
-			return fmt.Errorf("Cannot clone git repository to %s as the dir already exists. Maybe try 'cd %s' and re-run the 'jx boot' command?", repo, repo)
-		}
-
-		log.Logger().Infof("Cloning %s @ %s to %s\n", info(gitURL), info(gitRef), info(cloneDir))
-
-		err = os.MkdirAll(cloneDir, util.DefaultWritePermissions)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create directory: %s", cloneDir)
-		}
-
-		err = o.Git().Clone(gitURL, cloneDir)
-		if err != nil {
-			return errors.Wrapf(err, "failed to clone git URL %s to directory: %s", gitURL, cloneDir)
-		}
-		commitish, err := gits.FindTagForVersion(cloneDir, gitRef, o.Git())
-		if err != nil {
-			log.Logger().Debugf(errors.Wrapf(err, "finding tag for %s", gitRef).Error())
-		}
-
-		if commitish != "" {
-			err = o.Git().Reset(cloneDir, commitish, true)
-			if err != nil {
-				return errors.Wrapf(err, "setting HEAD to %s", commitish)
-			}
-		} else {
-			log.Logger().Infof("fetching branch %s", gitRef)
-			err = o.Git().FetchBranch(cloneDir, "origin", gitRef)
-			if err != nil {
-				return errors.Wrapf(err, "fetching branch %s", gitRef)
-			}
-
-			branchName := fmt.Sprintf("pr-%s", uuid.New().String())
-
-			err = o.Git().CreateBranchFrom(cloneDir, branchName, gitRef)
-			if err != nil {
-				return errors.Wrapf(err, "create branch %s from %s", branchName, gitRef)
-			}
-
-			err = o.Git().Checkout(cloneDir, branchName)
-			if err != nil {
-				return errors.Wrapf(err, "checkout branch %s", branchName)
-			}
-		}
-		o.Dir, err = filepath.Abs(cloneDir)
-		if err != nil {
-			return err
-		}
-
-		pipelineFile := filepath.Join(o.Dir, config.RequirementsConfigFileName)
-		pipelineFileExists, err := util.FileExists(pipelineFile)
-		if err != nil {
-			return err
-		}
-
-		if !pipelineFileExists {
-			return fmt.Errorf("The cloned repository %s does not include a Jenkins X Pipeline file at %s", gitURL, pipelineFile)
-		}
-	}
-
 	projectConfig, pipelineFile, err := config.LoadProjectConfig(o.Dir)
 	if err != nil {
 		return err
@@ -271,19 +186,6 @@ func (o *BootOptions) Run() error {
 
 	if err := o.overrideRequirements(gitURL); err != nil {
 		return errors.Wrap(err, "overwriting the default requirements")
-	}
-
-	requirements, requirementsFile, err := config.LoadRequirementsConfig(o.Dir)
-	if err != nil {
-		return errors.Wrap(err, "failed to load jx-requirements.yml file")
-	}
-
-	isBootClone, err = util.FileExists(requirementsFile)
-	if err != nil {
-		return err
-	}
-	if !isBootClone {
-		return fmt.Errorf("no requirements file %s are you sure you are running this command inside a GitOps clone?", requirementsFile)
 	}
 
 	// only update boot if the a GitRef has not been supplied
@@ -365,6 +267,108 @@ func (o *BootOptions) Run() error {
 	no.CommonOptions = o.CommonOptions
 	no.Args = []string{requirements.Cluster.Namespace}
 	return no.Run()
+}
+
+func (o *BootOptions) isGitRepo(dir string) bool {
+	_, _, err := gits.GetGitInfoFromDirectory(dir, o.Git())
+	if err == nil {
+		return true
+	}
+	return false
+}
+
+func (o *BootOptions) determineGitURLAndRef() (string, string) {
+	gitURL, gitRef, err := gits.GetGitInfoFromDirectory(o.Dir, o.Git())
+	if err != nil {
+		log.Logger().Info("Creating boot config with defaults, as not in an existing boot directory with a git repository.")
+		gitURL = config.DefaultBootRepository
+		gitRef = config.DefaultVersionsRef
+	}
+
+	if o.GitURL != "" {
+		log.Logger().Infof("GitURL provided, overriding the current value: %s with %s", util.ColorInfo(gitURL), util.ColorInfo(o.GitURL))
+		gitURL = o.GitURL
+	}
+
+	if o.GitRef != "" {
+		log.Logger().Infof("GitRef provided, overriding the current value: %s with %s", util.ColorInfo(gitRef), util.ColorInfo(o.GitRef))
+		gitRef = o.GitRef
+	}
+	return gitURL, gitRef
+}
+
+func (o *BootOptions) createBootClone(bootConfigGitURL string, bootConfigGitRef string, cloneDir string) (string, error) {
+	info := util.ColorInfo
+	log.Logger().Infof("No Jenkins X pipeline file %s or no jx boot requirements file %s found. You are not running this command from inside a "+
+		"Jenkins X Boot git clone", info(config.ProjectConfigFileName), info(config.RequirementsConfigFileName))
+
+	if !o.BatchMode {
+		log.Logger().Infof("To continue we will clone %s @ %s to %s", info(bootConfigGitURL), info(bootConfigGitRef), info(cloneDir))
+
+		help := "A git clone of a Jenkins X Boot source repository is required for 'jx boot'"
+		message := "Do you want to clone the Jenkins X Boot Git repository?"
+		if answer, err := util.Confirm(message, true, help, o.GetIOFileHandles()); err != nil {
+			return "", errors.Wrapf(err, "unable to process user input")
+		} else if !answer {
+			return "", fmt.Errorf("please run this command again inside a git clone from a Jenkins X Boot repository")
+		}
+	}
+
+	bootCloneExists, err := util.DirExists(cloneDir)
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to check whether directory %s exists", cloneDir)
+	}
+	if bootCloneExists {
+		return "", fmt.Errorf("cannot clone git repository to %s as the dir already exists", cloneDir)
+	}
+
+	log.Logger().Infof("Cloning %s @ %s to %s\n", info(bootConfigGitURL), info(bootConfigGitRef), info(cloneDir))
+
+	err = os.MkdirAll(cloneDir, util.DefaultWritePermissions)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create directory: %s", cloneDir)
+	}
+
+	err = o.Git().Clone(bootConfigGitURL, cloneDir)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to clone git URL %s to directory: %s", bootConfigGitURL, cloneDir)
+	}
+
+	commitish, err := gits.FindTagForVersion(cloneDir, bootConfigGitRef, o.Git())
+	if err != nil {
+		log.Logger().Debugf(errors.Wrapf(err, "finding tag for %s", bootConfigGitRef).Error())
+	}
+
+	if commitish != "" {
+		err = o.Git().Reset(cloneDir, commitish, true)
+		if err != nil {
+			return "", errors.Wrapf(err, "setting HEAD to %s", commitish)
+		}
+	} else {
+		log.Logger().Infof("fetching branch %s", bootConfigGitRef)
+		err = o.Git().FetchBranch(cloneDir, "origin", bootConfigGitRef)
+		if err != nil {
+			return "", errors.Wrapf(err, "fetching branch %s", bootConfigGitRef)
+		}
+
+		branchName := uuid.New().String()
+
+		err = o.Git().CreateBranchFrom(cloneDir, branchName, bootConfigGitRef)
+		if err != nil {
+			return "", errors.Wrapf(err, "create branch %s from %s", branchName, bootConfigGitRef)
+		}
+
+		err = o.Git().Checkout(cloneDir, branchName)
+		if err != nil {
+			return "", errors.Wrapf(err, "checkout branch %s", branchName)
+		}
+	}
+
+	cloneDir, err = filepath.Abs(cloneDir)
+	if err != nil {
+		return "", errors.Wrapf(err, "unable to determine absolute path for %s", cloneDir)
+	}
+	return cloneDir, nil
 }
 
 func (o *BootOptions) restoreFromDevEnvRepo() error {

--- a/pkg/cmd/boot/boot_integration_test.go
+++ b/pkg/cmd/boot/boot_integration_test.go
@@ -8,6 +8,12 @@ import (
 	"path/filepath"
 	"testing"
 
+	"github.com/jenkins-x/jx/pkg/cmd/clients"
+
+	cmd_mocks "github.com/jenkins-x/jx/pkg/cmd/clients/mocks"
+	"github.com/jenkins-x/jx/pkg/config"
+	"github.com/stretchr/testify/require"
+
 	"github.com/acarl005/stripansi"
 	"github.com/jenkins-x/jx/pkg/cmd/opts"
 	"github.com/jenkins-x/jx/pkg/gits"
@@ -23,6 +29,137 @@ const (
 
 	testFileName = "some-file"
 )
+
+func Test_createBootClone_into_empty_directory_succeeds(t *testing.T) {
+	log.SetOutput(ioutil.Discard)
+	bootDir, err := ioutil.TempDir("", "boot-test")
+	require.NoError(t, err)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
+
+	factory := cmd_mocks.NewMockFactory()
+	commonOpts := opts.NewCommonOptionsWithFactory(factory)
+	commonOpts.BatchMode = true
+	o := BootOptions{
+		CommonOptions: &commonOpts,
+		Dir:           bootDir,
+	}
+
+	repoPath := filepath.Join(bootDir, "jenkins-x-boot-config")
+	cloneDir, err := o.createBootClone(config.DefaultBootRepository, config.DefaultVersionsRef, repoPath)
+	assert.NoError(t, err)
+	assert.Contains(t, cloneDir, repoPath)
+}
+
+func Test_createBootClone_with_custom_ref_succeeds(t *testing.T) {
+	log.SetOutput(ioutil.Discard)
+	bootDir, err := ioutil.TempDir("", "boot-test")
+	require.NoError(t, err)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
+
+	factory := cmd_mocks.NewMockFactory()
+	commonOpts := opts.NewCommonOptionsWithFactory(factory)
+	commonOpts.BatchMode = true
+	o := BootOptions{
+		CommonOptions: &commonOpts,
+		Dir:           bootDir,
+	}
+
+	repoPath := filepath.Join(bootDir, "jenkins-x-boot-config")
+	cloneDir, err := o.createBootClone(config.DefaultBootRepository, "v1.0.70", repoPath)
+	assert.NoError(t, err)
+	assert.Contains(t, cloneDir, repoPath)
+
+	gitter := gits.NewGitCLI()
+	sha, err := gitter.GetLatestCommitSha(repoPath)
+	require.NoError(t, err)
+	assert.Equal(t, "424f97a9e1ad4a0b5e518cd492d2b16d8b8f6705", sha)
+}
+
+func Test_createBootClone_into_existing_git_repo_fails(t *testing.T) {
+	log.SetOutput(ioutil.Discard)
+	bootDir, err := ioutil.TempDir("", "boot-test")
+	require.NoError(t, err)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
+
+	factory := cmd_mocks.NewMockFactory()
+	commonOpts := opts.NewCommonOptionsWithFactory(factory)
+	commonOpts.BatchMode = true
+	o := BootOptions{
+		CommonOptions: &commonOpts,
+		Dir:           bootDir,
+	}
+
+	repoPath := filepath.Join(bootDir, "my-repo")
+	err = os.MkdirAll(repoPath, 0700)
+	require.NoError(t, err)
+	gitter := gits.NewGitCLI()
+	err = gitter.Init(repoPath)
+	require.NoError(t, err)
+
+	cloneDir, err := o.createBootClone(config.DefaultBootRepository, config.DefaultVersionsRef, repoPath)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "dir already exists")
+	assert.Empty(t, cloneDir)
+}
+
+func Test_createBootClone_into_existing_directory_fails(t *testing.T) {
+	log.SetOutput(ioutil.Discard)
+	bootDir, err := ioutil.TempDir("", "boot-test")
+	require.NoError(t, err)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
+
+	factory := cmd_mocks.NewMockFactory()
+	commonOpts := opts.NewCommonOptionsWithFactory(factory)
+	commonOpts.BatchMode = true
+	o := BootOptions{
+		CommonOptions: &commonOpts,
+		Dir:           bootDir,
+	}
+
+	cloneDir, err := o.createBootClone(config.DefaultBootRepository, config.DefaultVersionsRef, bootDir)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "dir already exists")
+	assert.Empty(t, cloneDir)
+}
+
+func Test_jx_boot_in_non_boot_repo_fails(t *testing.T) {
+	log.SetOutput(ioutil.Discard)
+	bootDir, err := ioutil.TempDir("", "boot-test")
+	require.NoError(t, err)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
+
+	commonOpts := opts.NewCommonOptionsWithFactory(clients.NewFactory())
+	commonOpts.BatchMode = true
+	o := BootOptions{
+		CommonOptions: &commonOpts,
+		Dir:           bootDir,
+	}
+
+	// make the tmp directory a git repo
+	gitter := gits.NewGitCLI()
+	err = gitter.Init(bootDir)
+	require.NoError(t, err)
+	err = gitter.SetRemoteURL(bootDir, "origin", "https://github.com/johndoe/jx.git")
+	require.NoError(t, err)
+	_, err = os.Create(filepath.Join(bootDir, "foo"))
+	require.NoError(t, err)
+	err = gitter.AddCommit(bootDir, "adding foo")
+	require.NoError(t, err)
+
+	err = o.Run()
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "trying to execute 'jx boot' from a non requirements repo")
+}
 
 func TestUpdateBootCloneIfOutOfDate_Conflicts(t *testing.T) {
 	gitter := gits.NewGitCLI()

--- a/pkg/cmd/boot/boot_test.go
+++ b/pkg/cmd/boot/boot_test.go
@@ -5,9 +5,13 @@ package boot
 import (
 	"io"
 	"io/ioutil"
+	"log"
 	"os"
 	"path/filepath"
 	"testing"
+
+	cmd_mocks "github.com/jenkins-x/jx/pkg/cmd/clients/mocks"
+	"github.com/jenkins-x/jx/pkg/gits"
 
 	"github.com/jenkins-x/jx/pkg/cmd/opts"
 	"github.com/jenkins-x/jx/pkg/config"
@@ -40,11 +44,15 @@ func TestDetermineGitRef_DefaultGitUrl(t *testing.T) {
 	o := TestBootOptions{}
 	bootDir, err := ioutil.TempDir("", "boot-test")
 	require.NoError(t, err)
-	defer os.RemoveAll(bootDir)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
 	o.setup(defaultBootRequirements, bootDir)
 
 	dir := o.createTmpRequirements(t)
-	defer os.RemoveAll(dir)
+	defer func() {
+		_ = os.RemoveAll(dir)
+	}()
 	requirements, _, err := config.LoadRequirementsConfig(dir)
 	require.NoError(t, err, "unable to load tmp jx-requirements")
 	resolver := &versionstream.VersionResolver{
@@ -62,11 +70,15 @@ func TestDetermineGitRef_GitURLNotInVersionStream(t *testing.T) {
 	o := TestBootOptions{}
 	bootDir, err := ioutil.TempDir("", "boot-test")
 	require.NoError(t, err)
-	defer os.RemoveAll(bootDir)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
 	o.setup(defaultBootRequirements, bootDir)
 
 	dir := o.createTmpRequirements(t)
-	defer os.RemoveAll(dir)
+	defer func() {
+		_ = os.RemoveAll(dir)
+	}()
 	requirements, _, err := config.LoadRequirementsConfig(dir)
 	require.NoError(t, err, "unable to load tmp jx-requirements")
 	resolver := &versionstream.VersionResolver{
@@ -79,12 +91,15 @@ func TestDetermineGitRef_GitURLNotInVersionStream(t *testing.T) {
 }
 
 func TestCloneDevEnvironment(t *testing.T) {
+	t.Parallel()
 
 	url := "https://github.com/jenkins-x/jenkins-x-boot-config"
 	o := TestBootOptions{}
 	bootDir, err := ioutil.TempDir("", "boot-test")
 	require.NoError(t, err)
-	defer os.RemoveAll(bootDir)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
 	o.setup(defaultBootRequirements, bootDir)
 
 	cloned, dir, err := o.cloneDevEnvironment(url)
@@ -100,7 +115,9 @@ func TestCloneDevEnvironmentIncorrectParam(t *testing.T) {
 	o := TestBootOptions{}
 	bootDir, err := ioutil.TempDir("", "boot-test")
 	require.NoError(t, err)
-	defer os.RemoveAll(bootDir)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
 	o.setup(defaultBootRequirements, bootDir)
 	url := "not-a-url"
 
@@ -108,6 +125,104 @@ func TestCloneDevEnvironmentIncorrectParam(t *testing.T) {
 	assert.NotNil(t, err, "error should not be nil")
 	assert.False(t, cloned)
 	assert.Empty(t, dir)
+}
+
+func Test_determineGitURLAndRef_uses_git_repo_settings(t *testing.T) {
+	t.Parallel()
+
+	log.SetOutput(ioutil.Discard)
+	bootDir, err := ioutil.TempDir("", "boot-test")
+	require.NoError(t, err)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
+
+	commonOpts := opts.NewCommonOptionsWithFactory(cmd_mocks.NewMockFactory())
+	commonOpts.BatchMode = true
+	o := BootOptions{
+		CommonOptions: &commonOpts,
+		Dir:           bootDir,
+	}
+
+	// make the tmp directory a git repo
+	gitter := gits.NewGitCLI()
+	err = gitter.Init(bootDir)
+	require.NoError(t, err)
+	err = gitter.SetRemoteURL(bootDir, "origin", "https://github.com/acme/jenkins-x-boot-config.git")
+	require.NoError(t, err)
+	_, err = os.Create(filepath.Join(bootDir, "foo"))
+	require.NoError(t, err)
+	err = gitter.AddCommit(bootDir, "adding foo")
+	require.NoError(t, err)
+	err = gitter.CreateBranch(bootDir, "foo")
+	require.NoError(t, err)
+	err = gitter.Checkout(bootDir, "foo")
+	require.NoError(t, err)
+
+	gitURL, gitRef := o.determineGitURLAndRef()
+	assert.Equal(t, "https://github.com/acme/jenkins-x-boot-config", gitURL)
+	assert.Equal(t, "foo", gitRef)
+}
+
+func Test_determineGitURLAndRef_defaults_to_jenkins_x_boot_config(t *testing.T) {
+	t.Parallel()
+
+	log.SetOutput(ioutil.Discard)
+	bootDir, err := ioutil.TempDir("", "boot-test")
+	require.NoError(t, err)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
+
+	commonOpts := opts.NewCommonOptionsWithFactory(cmd_mocks.NewMockFactory())
+	commonOpts.BatchMode = true
+	o := BootOptions{
+		CommonOptions: &commonOpts,
+		Dir:           bootDir,
+	}
+
+	gitURL, gitRef := o.determineGitURLAndRef()
+	assert.Equal(t, "https://github.com/jenkins-x/jenkins-x-boot-config.git", gitURL)
+	assert.Equal(t, "master", gitRef)
+}
+
+func Test_determineGitURLAndRef_explicit_provided_git_options_win(t *testing.T) {
+	t.Parallel()
+
+	log.SetOutput(ioutil.Discard)
+	bootDir, err := ioutil.TempDir("", "boot-test")
+	require.NoError(t, err)
+	defer func() {
+		_ = os.RemoveAll(bootDir)
+	}()
+
+	commonOpts := opts.NewCommonOptionsWithFactory(cmd_mocks.NewMockFactory())
+	commonOpts.BatchMode = true
+	o := BootOptions{
+		CommonOptions: &commonOpts,
+		Dir:           bootDir,
+		GitURL:        "https://github.com/johndoe/jenkins-x-boot-config.git",
+		GitRef:        "bar",
+	}
+
+	// make the tmp directory a git repo
+	gitter := gits.NewGitCLI()
+	err = gitter.Init(bootDir)
+	require.NoError(t, err)
+	err = gitter.SetRemoteURL(bootDir, "origin", "https://github.com/acme/jenkins-x-boot-config.git")
+	require.NoError(t, err)
+	_, err = os.Create(filepath.Join(bootDir, "foo"))
+	require.NoError(t, err)
+	err = gitter.AddCommit(bootDir, "adding foo")
+	require.NoError(t, err)
+	err = gitter.CreateBranch(bootDir, "foo")
+	require.NoError(t, err)
+	err = gitter.Checkout(bootDir, "foo")
+	require.NoError(t, err)
+
+	gitURL, gitRef := o.determineGitURLAndRef()
+	assert.Equal(t, "https://github.com/johndoe/jenkins-x-boot-config.git", gitURL)
+	assert.Equal(t, "bar", gitRef)
 }
 
 func (o *TestBootOptions) createTmpRequirements(t *testing.T) string {

